src/compiler/checker.ts

```diff
@@ -13349,23 +13349,60 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return result;
     }
 
+    function getFallbackExpandedLabels(restType: TupleTypeReference, restDeclaration: Declaration | undefined) {
+        if (!restDeclaration) {
+            return;
+        }
+        Debug.assertNode(restDeclaration, isParameter);
+        if (restDeclaration.name.kind !== SyntaxKind.ArrayBindingPattern) {
+            return;
+        }
+        const nameElements = restDeclaration.name.elements;
+        const elementFlags = restType.target.elementFlags;
+        const result: (__String | undefined)[] = arrayOf(elementFlags.length, () => undefined);

```

Comment:
On second thought, I don't know how I feel about this, but I'm also still not coming up with a better algorithm. I was probably too nitpicky before.

Reply:
I think this might not be equivalent. The `elementFlag & ElementFlags.Variable` case doesn't immediately "stop" the iteration and I think that the top condition (`seenDotDotDot || i >= nameElements.length`) doesn't cover this case either.

I pushed out your proposed code with a tweak to address the above. I'm not sure if I find it particularly more readable but I appreciate how it avoids an extra iteration so I'm fine with it.